### PR TITLE
Adds py lib and examples

### PR DESCRIPTION
**Pull request by jcooklin**
## Adds snap py plugin lib

Run tests 
- [x] Add shim between the user and the meta classes generated by PB
- [x] Ensure python3 support
- [x] Complete io documentation (https://intelsdi-x.github.io/snap-plugin-lib-py)
  - [x] module/class/method
  - [x] plugin authoring - collector
  - [x] plugin authoring - processor
  - [x] plugin authoring - publisher
  - [x] plugin packaging 

### How to run tests

1. install deps `pip install -r requirements.txt && pip install -r test-requirements.txt`
2. pytest (or tox)

tox

### Testing manually

I have a PR that I will open against Snap soon that will add support for plugin packages (aci based) that include relocated virtualenvs however until then you can test python plugins if snapd is run in an environment that has access to the plugin lib.

demo1

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 76e7b2a9fbf267d756989a52259e2566a7c68e15
  * Invalid type. Valid types are **fix, sdi-, maint**

 * Commit: 1540c019300752b6e32c93f04fe5cd4c03144cc7
  * Invalid type. Valid types are **fix, sdi-, maint**

 * Commit: 5d9e7147a9c0cf40fc79617bd2a8158803c76674
  * Invalid type. Valid types are **fix, sdi-, maint**

 * Commit: 9fc6ecd9b07a4e407aa64d3041ef6c02f9c8cf4d
  * Invalid type. Valid types are **fix, sdi-, maint**

 * Commit: 1cb65364bd88bc4776e0442cbcc774bddbe0a26d
  * Invalid type. Valid types are **fix, sdi-, maint**



---
This message was auto-generated by https://gitcop.com

File: snap_plugin/v1/bool_policy.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        for arg in args:
+            self._pb.rules[key].MergeFrom(arg._pb)
+
+
+class BoolRule(object):
+    """A configuration item with a integer value type.
+
+    Args:
+        default (:obj:`bool`): The default value
+        required (:obj:`bool`): Is the config item required
```
Comment (IRCody) [inline snap_plugin/v1/bool_policy.py]: Add a question mark here?

File: snap_plugin/v1/collector.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        avoided.  Only in rare circumstances should a plugin require
+        configuration data for it to load.
+
+        Args:
+            config (:obj:`snap_plugin.v1.ConfigMap`):
+                Provides configuration data.
+
+        Returns:
+            :obj:`list` of :obj:`snap_plugin.v1.Metric`:
+                List of collected metrics.
```
Comment (IRCody) [inline snap_plugin/v1/collector.py]: collectable metrics?

File: snap_plugin/v1/config_map.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# limitations under the License.
+
+from collections import MutableMapping
+from itertools import chain
+
+from .plugin_pb2 import ConfigMap as PbConfigMap
+
+
+class ConfigMap(MutableMapping):
+    """TODO (JC) add docstring
```
Comment (IRCody) [inline snap_plugin/v1/config_map.py]: .

File: snap_plugin/v1/get_metrictypes_arg.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .config_map import ConfigMap
+from .plugin_pb2 import GetMetricTypesArg as PbGetMetricTypesArg
+
+
+# class GetMetricTypesArg(with_metaclass(PbGetMetricTypesArg)):
```
Comment (IRCody) [inline snap_plugin/v1/get_metrictypes_arg.py]: .

File: snap_plugin/v1/integer_policy.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .plugin_pb2 import IntegerPolicy as PbIntegerPolicy
+from .plugin_pb2 import IntegerRule as PbIntegerRule
+
+
+class _IntegerPolicy(object):
```
Comment (IRCody) [inline snap_plugin/v1/integer_policy.py:25]: Do we want docstrings for the Policy classes also?

File: snap_plugin/v1/metrics_arg.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .plugin_pb2 import MetricsArg as PbMetricsArg
+
+
+class MetricsArg(object):
+    def __init__(self, *metrics):
+        # TODO(JC) need a docstring
```
Comment (IRCody) [inline snap_plugin/v1/metrics_arg.py]: .

File: snap_plugin/v1/metrics_arg.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        self._pb = PbMetricsArg(metrics=[m.pb for m in metrics])
+        # self._pb.metrics = [m.pb for m in metrics]
```
Comment (IRCody) [inline snap_plugin/v1/metrics_arg.py]: .

File: snap_plugin/v1/namespace.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                namespace element
+
+        Returns:
+            :py:class:`snap_plugin.v1.namespace.Namespace`
+        """
+        self._pb.add(Value=value)
+        return self
+
+    def pop(self, key=-1):
+        """Removes and retuns an the namespace element at a given index.
```
Comment (IRCody) [inline snap_plugin/v1/namespace.py]: Remove `an`?

File: snap_plugin/v1/plugin.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        self.cache_ttl = cache_ttl
+        self.rpc_type = rpc_type
+        self.rpc_version = rpc_version
+        self.unsecure = unsecure
+
+
+class Plugin(object):
+    """Abstract base class for Collector, Processor and Publisher plugins.
+
+    Plugin authoris shoud not inherit directly from this class rather they
```
Comment (IRCody) [inline snap_plugin/v1/plugin.py:133]: `authors`

File: snap_plugin/v1/plugin.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+            - Finding an available port
+            - Starting the gRPC server
+            - Printing to STDOUT data (JSON) for handshaking with snapd
+        """
+        LOG.debug("plugin start called..")
+        # find available port for grpc server
+        with closing(socket.socket()) as _socket:
+            _socket.bind(('', 0))
+            self._port = _socket.getsockname()[1]
+            self.server.add_insecure_port('[::]:{!s}'.format(self._port))
```
Comment (IRCody) [inline snap_plugin/v1/plugin.py]: Adding a note here b/c I had issues with errors around finding open port for grpc and I want to come back to this.

File: snap_plugin/v1/processor_proxy.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        LOG.debug("Process called")
+        try:
+            metrics = self.plugin.process(
+                [Metric(pb=m) for m in request.Metrics],
+                ConfigMap(pb=request.Config)
+            )
+            return MetricsReply(metrics=[m.pb for m in metrics])
+        except Exception as err:
+            return MetricsReply(metrics=[], error=err.message)
+        # return self.plugin.process(request.Metrics, request.Config)
```
Comment (IRCody) [inline snap_plugin/v1/processor_proxy.py]: .

File: snap_plugin/v1/publisher.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        by the plugin which extends :obj:`snap_plugin.v1.Publisher`.
+
+        This method is called by the snapd deamon during the publish phase
+        of a Snap workflow.
+
+        Args:
+            metrics (:obj:`list` of `snap_plugin.v1.Metric`):
+                List of metrics to be published.
+            config (`snap_plugin.v1.ConfigMap`):
+                map of config values.
```
Comment (IRCody) [inline snap_plugin/v1/publisher.py]: Should map be capitalized? Also I wonder if for python we should refer to it as a dictionary?

File: examples/collector/rand.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+import random
+import time
+
+import snap_plugin.v1 as snap
+
+LOG = logging.getLogger(__name__)
+
+
+class Rand(snap.Collector):
+    """ TODO (JC) add class docstring """
```
Comment (IRCody) [inline examples/collector/rand.py]: .

File: examples/processor/tag.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+import logging
+
+import snap_plugin.v1 as snap
+
+LOG = logging.getLogger(__name__)
+
+
+class Tag(snap.Processor):
+    """ TODO (JC) add class docstring"""
```
Comment (IRCody) [inline examples/processor/tag.py]: .

File: docs/source/plugin_authoring/processor.rst [hunk 15]
```diff
@@ -0,0 +1,7 @@
+.. _Processor-label:
+
+Processor
+---------
+
+Below you will find details that cover writing a snap processor plugin in
+Python.
```
Comment (IRCody) [inline docs/source/plugin_authoring/processor.rst]: todo?

File: docs/source/plugin_authoring/publisher.rst [hunk 16]
```diff
@@ -0,0 +1,7 @@
+.. _publisher-label:
+
+Publisher
+---------
+
+Below you will find details that cover writing a Snap publisher plugin in
+Python.
```
Comment (IRCody) [inline docs/source/plugin_authoring/publisher.rst]: todo?

**Review by IRCody — CHANGES_REQUESTED** [review]
Did a quick(ish) first pass through the commits. Going to try to use the library next and see if there are any issues.

File: snap_plugin/v1/plugin.py (hunk 10 above)
Comment (IRCody) [reply snap_plugin/v1/plugin.py]: I got it to work by using setsockopt(SOL_SOCKET, SO_REUSEADDR, 1) on _socket. Also needed to switch to using `127.0.0.1` instead of `[::]` since that will break if ipv6 is enabled.

**Comment by IRCody** [comment]
The latest changes around the port/ipv4vs6 fix the issues I was having. I'm not able to load plugins/ run tasks that use those plugins without issues.

File: snap_plugin/v1/plugin.py [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+    def start_plugin(self):
+        """Starts the Plugin
+
+        Starting a plugin includes:
+            - Finding an available port
+            - Starting the gRPC server
+            - Printing to STDOUT data (JSON) for handshaking with snapd
+        """
+        LOG.debug("plugin start called..")
```
Comment (kindermoumoute) [inline snap_plugin/v1/plugin.py:189]: Is it possible to handle [log level and timeout duration](https://github.com/intelsdi-x/snap/blob/master/control/plugin/plugin.go#L135-L138) from stdin?

File: snap_plugin/v1/metric.py [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+            elif self._pb.HasField("int64_data"):
+                self._data_type = int
+            elif self._pb.HasField("float64_data"):
+                self._data_type = float
+            elif self._pb.HasField("string_data"):
+                self._data_type = str
+            elif self._pb.HasField("bool_data"):
+                self._data_type = bool
+            elif self._pb.HasField("bytes_data"):
+                self._data_type = bytes
```
Comment (kindermoumoute) [inline snap_plugin/v1/metric.py:69]: Does this list need to be extended for https://github.com/intelsdi-x/snap/issues/1250 ?
(I think @candysmurf is working on that)

File: snap_plugin/v1/metric.py [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+                or :obj:`basestring` or :obj:`bytes`)
+
+        Returns:
+            value
+
+        Raises:
+            :obj:`TypeError`
+
+        """
+        types = {
```
Comment (kindermoumoute) [inline snap_plugin/v1/metric.py:232]: Same than above.

File: snap_plugin/v1/plugin_pb2.py [hunk 20]
```diff
@@ -0,0 +1,2293 @@
...
+      options=None),
+    _descriptor.FieldDescriptor(
+      name='Description', full_name='rpc.Metric.Description', index=7,
+      number=8, type=9, cpp_type=9, label=1,
+      has_default_value=False, default_value=_b("").decode('utf-8'),
+      message_type=None, enum_type=None, containing_type=None,
+      is_extension=False, extension_scope=None,
+      options=None),
+    _descriptor.FieldDescriptor(
+      name='string_data', full_name='rpc.Metric.string_data', index=8,
```
Comment (kindermoumoute) [inline snap_plugin/v1/plugin_pb2.py:306]: Same than above.

File: docs/source/plugin_authoring/packaging.rst [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+#########
+Packaging
+#########
+
+Below you find details that cover packaging a Python based Snap plugin.
+Packaging a python plugin depends on the following tools being installed and 
+appropriately configured.
+
+- `acbuld <https://github.com/containers/build>`_
```
Comment (kindermoumoute) [inline docs/source/plugin_authoring/packaging.rst]: *acbuild

File: snap_plugin/v1/bool_policy.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+class _BoolPolicy(object):
+    def __init__(self, key, *args, **kwargs):
+        self._pb = kwargs.get("pb", PbBoolPolicy())
+        for arg in args:
+            self._pb.rules[key].MergeFrom(arg._pb)
+
+
+class BoolRule(object):
+    """A configuration item with a integer value type.
```
Comment (marcin-krolik) [inline snap_plugin/v1/bool_policy.py:39]: should it be "bool value type"?

File: snap_plugin/v1/namespace.py [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+class Namespace(object):
+    """Namespace of a metric.
+
+    Args:
+        elements (:obj:`list` of :py:class:`~snap_plugin.v1.namespace_element.NamespaceElement`):
+            namespace elements
+
+    """
+
+    def __init__(self, pb, *elements):
```
Comment (marcin-krolik) [inline snap_plugin/v1/namespace.py:35]: Most metrics have static NamespaceElements only, I think it would be beneficial to add possibility to instantiate Namespace from a list of strings.

File: snap_plugin/v1/metric.py [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+        version (:obj:`int`): metric version
+        tags (:obj:`dict`): metric tags (key/value pairs)
+        config (:py:class:`~snap_plugin.v1.config_map.ConfigMap`): config for
+            metric (key/value pairs)
+        timestamp (:obj:`float`): metric timestamp (see time.time())
+        unit (:obj:`str`): metric unit
+        description (:obj:`str`): metric description
+
+    """
+    def __init__(self, namespace=[], version=0, tags={}, config={},
```
Comment (marcin-krolik) [inline snap_plugin/v1/metric.py]: Similarly here, I think giving plugin author ability to build metric with Namespace instance or list of strings would be much easier. Providing list of NamespaceElements in Metric constructor seems little bit too verbose and complicated from the perspective of plugin author. What is your opinion?

File: snap_plugin/v1/bool_policy.py (hunk 22 above)
Comment (jcooklin) [reply snap_plugin/v1/bool_policy.py:39]: No, not IMO.

File: snap_plugin/v1/namespace.py (hunk 23 above)
Comment (jcooklin) [reply snap_plugin/v1/namespace.py:35]: 👍 Done.

File: snap_plugin/v1/metric.py (hunk 24 above)
Comment (jcooklin) [reply snap_plugin/v1/metric.py]: Done.

File: snap_plugin/v1/metric.py (hunk 18 above)
Comment (jcooklin) [reply snap_plugin/v1/metric.py:69]: Once the plugin.proto in snap includes the new value fields we'll make the changes here.

File: snap_plugin/v1/integer_policy.py (hunk 5 above)
Comment (jcooklin) [reply snap_plugin/v1/integer_policy.py:25]: It's a private class that isn't exposed to the user and doesn't appear in the api docs.  The rules are documented and is the part of the ConfigPolicy that plugin authors will be exposed to.  Creating rules of a given type will automatically wrap them in the appropriate policy.  This all being said I'm going to add a brief doc string to all of the policies for those who might support the lib in the future.

Comment (jcooklin) [reply snap_plugin/v1/integer_policy.py:25]: example:

```
snap.ConfigPolicy(
            [
                ("random"),
                [
                    (
                        "int_max",
                        snap.IntegerRule(default=100, minimum=1, maximum=10000)
                    ),
                    (
                        "int_min",
                        snap.IntegerRule(default=0, minimum=0)
                    )
                ]
            ]
        )
```

File: docs/source/plugin_authoring/packaging.rst.orig [hunk 25]
```diff
@@ -0,0 +1,59 @@
+.. _packaging-label:
```
Comment (andrzej-k) [inline docs/source/plugin_authoring/packaging.rst.orig]: Is this file needed?

**Review by IRCody — APPROVED** [review]
LGTM
